xBridge battery display on Home preference

jstevensog opened the pull request:
Added a preference (and slightly re-ordered the preferences around Data Source for better flow) to allow people to turn on/off the xBridge battery display in Home. Default is on.
This allows those NOT using a voltage divider circuit (always 0% - Waiting for Packet in red), or those powering the bridge from a phone (always 100%) to opt to turn this display off.

I believe this is better than simply hiding the battery display if it is 0%, as it covers both cases.
Cheers

jstevensog committed:
Remove extra BgGraphBuilder creation in bdDelta().

jstevensog committed:
Preference for xBridge battery display in home.

Added a preference to display/not display the xBridge battery capacity in
home.
This will allow those NOT using a voltage divider, or those powering their
bridge directly from a phone to choose NOT to display the bridge battery
level on home.

AdrianLxM commented:
From pure code review it looks good to me.  :)

JohanDegraeve commented:
I will test it

JohanDegraeve commented:
I don't know exactly how to propose a pull request to this pull request so I write my comments : 

if (bridgeBattery == 0) {
                dexbridgeBattery.setText("Waiting for packet");
==> i propose to change to
**if (bridgeBattery == 0) {
                dexbridgeBattery.setText("xBridge battery level unknown, waiting for packet");**
"waiting for packet" looks rather confusing to me for the user. A user will not know this is linked to the setting "Display Bridge Battery"

and also keep the text white if bridgebattery == 0
so after the check on bridgebattery < 25 add
**if (bridgeBattery == 0) dexbridgeBattery.setTextColor(Color.WHITE);**

by the way i think the color can never be yellow. If value is eg 30, it will first be set to yellow, then to green

jstevensog commented:
Great feedback Johan.
I will have a go at incorporating all of that in the branch, which should
update the PR automatically (I hope).
Lunch time isn't to far away, so I might get to play with it then.
Cheers

On Thu, Dec 10, 2015 at 10:49 AM, Johan Degraeve notifications@github.com
wrote:

> I don't know exactly how to propose a pull request to this pull request so
> I write my comments :
> 
> if (bridgeBattery == 0) {
> dexbridgeBattery.setText("Waiting for packet");
> ==> i propose to change to
> 
> _if (bridgeBattery == 0) { dexbridgeBattery.setText("xBridge battery level
> unknown, waiting for packet");_
> "waiting for packet" looks rather confusing to me for the user. A user
> will not know this is linked to the setting "Display Bridge Battery"
> 
> and also keep the text white if bridgebattery == 0
> so after the check on bridgebattery < 25 add
> _if (bridgeBattery == 0) dexbridgeBattery.setTextColor(Color.WHITE);_
> 
> by the way i think the color can never be yellow. If value is eg 30, it
> will first be set to yellow, then to green
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/209#issuecomment-163439675
> .

## 

John Stevens
"You are how you live, not what you have."

jstevensog commented:
Thinking about it some more, the "Waiting for Packet" will only ever appear
once.  Since Stephen moved the storage of the battery into a preference, it
is only ever 0 once.  Might need to rethink the whole thing.  Like, write
some code to set it to 0 if a packet hasn't been seen for >15 minutes.
Cheers

On Thu, Dec 10, 2015 at 10:49 AM, Johan Degraeve notifications@github.com
wrote:

> I don't know exactly how to propose a pull request to this pull request so
> I write my comments :
> 
> if (bridgeBattery == 0) {
> dexbridgeBattery.setText("Waiting for packet");
> ==> i propose to change to
> 
> _if (bridgeBattery == 0) { dexbridgeBattery.setText("xBridge battery level
> unknown, waiting for packet");_
> "waiting for packet" looks rather confusing to me for the user. A user
> will not know this is linked to the setting "Display Bridge Battery"
> 
> and also keep the text white if bridgebattery == 0
> so after the check on bridgebattery < 25 add
> _if (bridgeBattery == 0) dexbridgeBattery.setTextColor(Color.WHITE);_
> 
> by the way i think the color can never be yellow. If value is eg 30, it
> will first be set to yellow, then to green
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/209#issuecomment-163439675
> .

## 

John Stevens
"You are how you live, not what you have."

jstevensog committed:
Modified per Johan Degraeve's comments.

Minor changes to the logic around showing the Bridge Battery.

jstevensog committed:
Fixed Typo.

jstevensog committed:
Altered wording when xBridge battery is unknown.

JohanDegraeve commented:
with an xDrip that does not have the resistors, but with the xbridge2.wxl software installed, and configured on the app as an xbridge, if I enable the setting now "Disaplay Bridge Battery", then the red warning stays there for ever.

Personally i would makeit even more simply : as long as now bridge battery level > 0 is received, don't show the message at all, and don't show the setting at all.
As soon as a battery level > 0 is received, then show the setting (default checked), and then also show the message on the home screen according to the value of the setting.

Like this : xdrip users (ie with resistors but with xbridge2 software) will never notice, they don't need the feature.
xbridge users (ie with resistors) they will be able to use the feature

AdrianLxM commented:
@JohanDegraeve, how would you reset that?
Every time a new device is selected after a BT-Scan and every time the collection method is changed to xBridge?
That would be a logic that does not depend on the current state but on transitions from the state before (that you can no longer see on the interface) and therefore harder to support on facebook than a simple setting.

jstevensog commented:
Johan, I'll try and fix my logic.  I did it in a rush, and it seemed fine when I tried it.  That would be simplest I think.  There are people with the xDrip bridge running xBridge2 with the xDrip voltage divider resistors, which xBridge2 automatically detects and supports.

I am not sure your logic can be easily implemented to deal with all cases.
Cheers

---- Johan Degraeve wrote ----

> with an xDrip that does not have the resistors, but with the xbridge2.wxl software installed, and configured on the app as an xbridge, if I enable the setting now "Disaplay Bridge Battery", then the red warning stays there for ever.
> 
> Personally i would makeit even more simply : as long as now bridge battery level > 0 is received, don't show the message at all, and don't show the setting at all.
> As soon as a battery level > 0 is received, then show the setting (default checked), and then also show the message on the home screen according to the value of the setting.
> 
> Like this : xdrip users (ie with resistors but with xbridge2 software) will never notice, they don't need the feature.
> xbridge users (ie with resistors) they will be able to use the feature
> 
> —
> Reply to this email directly or view it on GitHub.￼

jstevensog committed:
Clear dexbridgeBattery string in Home.java

When xBridge is NOT the selected data source, or display_bridge_battery
preference is false.

JohanDegraeve commented:
I'll wait for the next update from John

JohanDegraeve commented on app/src/main/java/com/eveningoutpost/dexdrip/Home.java:
```diff
@@ -375,19 +375,22 @@ private void displayCurrentInfo() {
-            if (bridgeBattery < 25) dexbridgeBattery.setTextColor(Color.RED);
+            if (bridgeBattery < 50 && bridgeBattery >25) dexbridgeBattery.setTextColor(Color.YELLOW);
+            if (bridgeBattery <= 25) dexbridgeBattery.setTextColor(Color.RED);
             else dexbridgeBattery.setTextColor(Color.GREEN);
```
note that it wil still be reset to green whenver > 25, it will never be yellow
and also when bridgeBattery = 0, it's again set to Red.

jstevensog replied:
You are right, but I had not noticed that before.  Fixed the logic now I
believe.
I cannot see a reason why you would always have the red Waiting for Packet
message showing permanently though.  I will try and run it up on my test
phone and see if I can reproduce.
Cheers

On Fri, Dec 11, 2015 at 9:27 AM, Johan Degraeve notifications@github.com
wrote:

> In app/src/main/java/com/eveningoutpost/dexdrip/Home.java
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/209#discussion_r47296192
> :
> 
> > ```
> >          }
> > ```
> > -            if (bridgeBattery < 50) dexbridgeBattery.setTextColor(Color.YELLOW);
> > -            if (bridgeBattery < 25) dexbridgeBattery.setTextColor(Color.RED);
> > -            if (bridgeBattery < 50 && bridgeBattery >25) dexbridgeBattery.setTextColor(Color.YELLOW);
> > -            if (bridgeBattery <= 25) dexbridgeBattery.setTextColor(Color.RED);
> >            else dexbridgeBattery.setTextColor(Color.GREEN);
> 
> note that it wil still be reset to green whenver > 25, it will never be
> yellow
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/209/files#r47296192
> .

## 

John Stevens
"You are how you live, not what you have."

jstevensog committed:
Corrected Home dexBattery logic for colours.

As pointed out by Johan Degraeve, yellow would automatically have been
reset to Green.  So reordered the logic.
Also changed the Red limit to 30 to match the watch face.  I find that
once the bridge hits 30% battery, it rapidly depletes.

JohanDegraeve commented:
note that i use an xdrip without resistors but with xbridge2 software installed. I think it's normal that I see the red message because battery level stays 0

jstevensog commented:
Yep, I understood that.
xDrip bridge also has voltage divider resistors, and some use them with
their own modified xdrip-wixel code.
The xBridge2 code detects if the circuit is xBridge2 or xdrip-wixel and
sets up the voltage calculation of 0-100% appropriately.  It also "learns"
the limits of the individual circuit.  This is very rudimentary and has
hard limits on both in case something goes very awry.
xBridge2 voltage divider is lower current impact (better battery life), so
if you were ever going to add this to your setup, and are sticking with the
xdrip-wixel circuit, use the voltage divider resistors for xdrip-wixel, not
xBridge2.

Cheers

On Fri, Dec 11, 2015 at 10:00 AM, Johan Degraeve notifications@github.com
wrote:

> note that i use an xdrip without resistors but with xbridge2 software
> installed. I think it's normal that I see the red message because battery
> level stays 0
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/StephenBlackWasAlreadyTaken/xDrip-Experimental/pull/209#issuecomment-163779934
> .

## 

John Stevens
"You are how you live, not what you have."

jstevensog committed:
Clear bridge_battery preference on start.

If dexBridgeWixel is the data source, reset the bridge_battery preference
to 0.  That way people know no packet has been received yet, so don't
trust the battery capacity reading.
Probably overkill, but relevant.
